apps/cdn/src/cache.rs

```diff
@@ -0,0 +1,222 @@
+use std::{
+	hash::Hash,
+	str::FromStr,
+	sync::{atomic::AtomicUsize, Arc},
+};
+
+use reqwest::header::HeaderValue;
+use scc::{ebr::Guard, TreeIndex};
+use scuffle_foundations::http::server::{
+	axum::http::header::{self, HeaderMap, HeaderName},
+	axum::http::{Response, StatusCode},
+	stream::{Body, IntoResponse},
+};
+
+use crate::global::Global;
+
+const ONE_DAY: std::time::Duration = std::time::Duration::from_secs(60 * 60 * 24);
+
+pub struct Cache {
+	inner: moka::future::Cache<CacheKey, CachedResponse>,
+	path_meta: Arc<TreeIndex<String, Arc<PathMeta>>>,
+	http_client: reqwest::Client,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum CacheError {
+	#[error("failed to get cdn file: {0}")]
+	Reqwest(#[from] reqwest::Error),
+	#[error("failed to update path meta index")]
+	PathMetaUpdate,
+}
+
+impl Cache {
+	pub fn new(capacity: u64) -> Self {
+		let path_meta = Arc::new(TreeIndex::new());
+		let pm = Arc::clone(&path_meta);
+
+		Self {
+			inner: moka::future::Cache::builder()
+				.async_eviction_listener(move |key: Arc<CacheKey>, _, _| {
+					let path_meta = Arc::clone(&pm);
+
+					Box::pin(async move {
+						path_meta.remove_async(&key.path).await;
+					})
+				})
+				.expire_after(CacheExpiry)
+				.weigher(|_, v: &CachedResponse| v.body.len() as u32)
+				.max_capacity(capacity)
+				.build(),
+			path_meta,
+			http_client: reqwest::Client::new(),
+		}
+	}
+
+	pub async fn handle_request(
+		&self,
+		global: &Arc<Global>,
+		key: String,
+		mut headers: HeaderMap,
+	) -> Result<CachedResponse, CacheError> {
```

Comment:
This does not do [request coalescing](https://support.bunny.net/hc/en-us/articles/6762047083922-Understanding-Request-Coalescing). Perhaps have some mechanism to do that otherwise hot resources might generate 1000s of request to the upstream while we are waiting for it to be cached.